test/controllers/products_controller_test.rb

```diff
@@ -0,0 +1,261 @@
+        product_id = Product.last.id + 1
+
+        get edit_product_path(product_id)
+
+        must_respond_with :not_found
+      end
+    end
+
+    describe 'update' do
+      it 'updates an exsisting product with valid data' do
+        product = Product.first
+        product_data = product.attributes
+        product_data[:name] = "some updated name"
+
+        # Assumptions
+        product.assign_attributes(product_data)
+        product.must_be :valid?
+
+        # Act
+        patch product_path(product), params: { product: product_data }
+
+        # Assert
+        must_redirect_to product_path(product)
+
+        product.reload
+        product.name.must_equal product_data[:name]
+      end
+
+      it 'sends a bad request for invalid data' do
+        product = Product.first
+        product_data = product.attributes
+        product_data[:name] = ""
+
+        # Assumptions
+        product.assign_attributes(product_data)
+        product.wont_be :valid?
+
+        # Act
+        patch product_path(product), params: { product: product_data }
+
+        # Assert
+        must_respond_with :bad_request
+
+        product.reload
+        product.name.wont_equal product_data[:name]
+      end
+    end
+
+    describe 'retire' do
+      before do
+        login(Merchant.first)
+        @product = Product.first
+        @product.merchant_id = @merchant.id
+      end
+
+      it 'can successfully retire a product that still has stock' do
+        @product.stock = 4
+
+        patch retire_path(@product)
+        @product.reload
+        @product.wont_be :visible
+        must_respond_with :redirect
+        flash[:status].must_equal :success
+      end
+
+      it "retires an item that stock has reached 0" do
+        @product.stock = 0
+        @product.visible = false
+        @product.save
+
+        patch retire_path(@product)
+        @product.reload
+        @product.wont_be :visible
+        must_respond_with :redirect
+        flash[:status].must_equal :failure
+      end
+
+      it 'will not retire an item that is already retired and will respond with result_text if the product cannot be retired' do
+        @product.visible = false
+        @product.stock = 0
+        @product.save
+        patch retire_path(@product)
+        @product.reload
+        @product.wont_be :visible
+        must_respond_with :redirect
+        flash[:status].must_equal :failure
+      end
+
+      it 'unretires an item thats stock is not 0' do
+        @product.visible = false
+        @product.stock = 5
+        @product.save
+        patch retire_path(@product)
+        @product.reload
+        @product.must_be :visible
+        must_respond_with :redirect
+        flash[:status].must_equal :success
+      end
+    end
+  end
+  # These are tests for a guest user. No ability to alter or create products. Only show and index.
+
+
+  describe 'guest user' do
+    it 'rejects requests for new product form' do
```

Comment:
You have tests for a guest and logged in user, but your authorization story is more nuanced. What if you're logged in as merchant A, and try to edit/update/delete a product sold by merchant B? Clearly you shouldn't be able to, but your tests must exercise this behavior.